crates/subspace-archiving/src/piece_reconstructor.rs

```diff
@@ -0,0 +1,176 @@
+#[derive(Debug, Clone)]
+pub struct PiecesReconstructor {
+    /// Configuration parameter defining the size of one record (data in one piece excluding witness
+    /// size)
+    record_size: u32,
+    /// Configuration parameter defining the size of one recorded history segment
+    segment_size: u32,
+    /// Erasure coding data structure
+    reed_solomon: ReedSolomon,
+    /// KZG instance
+    kzg: Kzg,
+}
+
+impl PiecesReconstructor {
+    fn shards_count(segment_size: u32, record_size: u32) -> u32 {
+        segment_size / record_size
+    }
+
+    fn data_shards(&self) -> u32 {
+        Self::shards_count(self.segment_size, self.record_size)
+    }
+
+    fn parity_shards(&self) -> u32 {
+        self.data_shards()
+    }
+
+    pub fn new(
+        record_size: u32,
+        segment_size: u32,
+        kzg: Kzg,
+    ) -> Result<Self, ReconstructorInstantiationError> {
+        if segment_size <= record_size {
+            return Err(ReconstructorInstantiationError::SegmentSizeTooSmall);
+        }
+        if segment_size % record_size != 0 {
+            return Err(ReconstructorInstantiationError::SegmentSizesNotMultipleOfRecordSize);
+        }
+
+        let data_shards = Self::shards_count(segment_size, record_size);
+        let parity_shards = data_shards;
+        let reed_solomon = ReedSolomon::new(data_shards as usize, parity_shards as usize)
+            .expect("ReedSolomon must always be correctly instantiated");
+
+        Ok(Self {
+            record_size,
+            segment_size,
+            reed_solomon,
+            kzg,
+        })
+    }
+
+    /// Returns all the pieces for a segment using given set of pieces of a segment of the archived
+    /// history (any half of all pieces are required to be present, the rest will be recovered
+    /// automatically due to use of erasure coding if needed).
+    pub fn retrieve_pieces(
+        &self,
+        segment_pieces: &[Option<Piece>],
+    ) -> Result<Vec<Piece>, PieceReconstructorError> {
+        // If not all data pieces are available, need to reconstruct data shards using erasure
+        // coding.
+        let mut shards = segment_pieces
+            .iter()
+            .map(|maybe_piece| maybe_piece.as_ref().map(utils::slice_to_arrays))
+            .collect::<Vec<_>>();
+
+        self.reed_solomon
+            .reconstruct(&mut shards)
+            .map_err(PieceReconstructorError::DataShardsReconstruction)?;
+
+        let reconstructed_record_shards = shards
+            .iter()
+            .map(|maybe_shard| {
+                maybe_shard
+                    .as_ref()
+                    .map(|shard| {
+                        let bytes = shard.iter().fold(Vec::new(), |mut acc, shard_part| {
+                            acc.extend_from_slice(shard_part);
+
+                            acc
+                        });
+
+                        let record = &bytes[0..self.record_size as usize];
+
+                        record.to_vec()
```

Comment:
`.fold()` is almost always a less intuitive API. Also this looks very wasteful:
1) You know the eventual size precisely, yet you create `Vec::new()` rather than `Vec::with_capacity()`
2) Then you slice the record out of it rather than either shrinking the original vector you're about to throw away anyway (and wasting some RAM in the process) or allocating space just for record to begin with and don't even copy witness from `shard` into the resulting record at all

These reconstruction APIs are potentially performance-sensitive and should be written carefully. These are not tests that should just work, these need to be carefully designed and be efficient, especially with low-hanging fruits like these.

Reply:
Just in case. You can use [`flatten`](https://doc.rust-lang.org/std/primitive.slice.html#method.flatten) method on a slice (unfortunately it requires nightly feature), so you can do the following:
```rust
shard.flatten()[..self.record_size as usize].to_owned()
```